src/librustc_mir/transform/const_prop.rs

```diff
@@ -659,9 +658,58 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
             return None;
         }
 
+        self.eval_rvalue_with_identities(rvalue, place)
+    }
+
+    // Attempt to use albegraic identities to eliminate constant expressions
+    fn eval_rvalue_with_identities(
+        &mut self,
+        rvalue: &Rvalue<'tcx>,
+        place: Place<'tcx>,
+    ) -> Option<()> {
         self.use_ecx(|this| {
-            trace!("calling eval_rvalue_into_place(rvalue = {:?}, place = {:?})", rvalue, place);
-            this.ecx.eval_rvalue_into_place(rvalue, place)?;
+            match rvalue {
+                Rvalue::BinaryOp(op, left, right) | Rvalue::CheckedBinaryOp(op, left, right) => {
+                    let l = this.ecx.eval_operand(left, None);
+                    let r = this.ecx.eval_operand(right, None);
+
+                    if let (Ok(_), Ok(_)) = (&l, &r) {
+                        return Ok(this.ecx.eval_rvalue_into_place(rvalue, place)?);
+                    }
+
+                    let const_arg = this.ecx.read_immediate(l.or(r)?)?;
+                    let dest = this.ecx.eval_place(place)?;
+
+                    match op {
+                        BinOp::BitAnd => {
+                            if const_arg.to_scalar()?.to_bool()? == false {
```

Comment:
this could be generalized to all integers by checking `.to_bits_or_ptr(size, &this.ecx).ok()? == 0` similarly for the `BitOr`, but there you need to compare against `mir::truncate(u128::max_value(), size)`

In both cases `size` is `l.or(r)?.layout.size`

Reply:
👍

Reply:
Hmm the problem is that in the case of booleans it's truncated to a whole byte rather than a single bit so the max value is 255 :/ I guess I'll just check for bools and all other ints seperately.

Reply:
oh :( yea that's unfortunate